File: frontend/apps/ios/Pantopus/Features/Scheduling/Business/RoundRobin/RoundRobinAssignmentViewModel.swift
//
//  RoundRobinAssignmentViewModel.swift
//  Pantopus
//
//  G1 Round-Robin Assignment (Stream I13) — bottom sheet hung off a business
//  service. Picks which members take bookings and the fairness rule. Saving
//  REPLACES the whole assignee set via PUT /event-types/:id/assignees
//  (INVALID_ASSIGNEE surfaced). Round-robin uses weight (Balanced) / priority
//  (Priority order) / equal (Strict). Matches `roundrobin-frames.jsx`
//  (default / loading / none-selected / single-member).
//
//  Backend note: there is no dedicated "rule" column — Balanced/Priority/Strict
//  are expressed through the assignees' weight/priority and inferred on reload.
//  The encoding is a lossless sentinel scheme so the chosen rule survives a
//  save→reload round-trip even when every value is at its default:
//    · strict   → weight 1, priority 0            (the neutral shape)
//    · priority → weight 1, priority index+1      (1-based rank; >0 marks the rule)
//    · balanced → user weights, priority -1       (<0 marks the rule; weights carry the data)
//  All three are behavior-neutral for the backend's host pick
//  (bookingService.pickRoundRobinHost sorts by assigned_count, then priority
//  DESC — equal priorities don't reorder, and +1 on every rank preserves the
//  relative order the old 0-based encoding produced). `weight` is not read by
//  the pick at all. Joi (weight min 0, priority any integer) and the DB column
//  (plain integer, no CHECK) both accept the sentinel values.
//

import Foundation
import Observation

@Observable
@MainActor
final class RoundRobinAssignmentViewModel {
    enum Rule: String, CaseIterable, Equatable { case balanced, priority, strict }

    enum Phase: Equatable { case loading, ready, error(String) }

    struct Pick: Identifiable, Hashable {
        let id: String
        let name: String
        let avatarURL: String?
        let role: String?
        var checked: Bool
        var weight: Int
    }

    // MARK: Inputs

    let owner: SchedulingOwner
    let eventTypeId: String
    private let client: SchedulingClient

    // MARK: State

    private(set) var phase: Phase = .loading
    var selectedRule: Rule = .balanced
    private(set) var picks: [Pick] = []
    private(set) var isSaving = false
    private(set) var saveError: String?

    // MARK: Derived

    var checkedCount: Int {
        picks.filter(\.checked).count
    }

    var doneDisabled: Bool {
        checkedCount == 0 || isSaving
    }

    var isSingleMember: Bool {
        checkedCount == 1
    }

    var firstCheckedName: String? {
        picks.first(where: \.checked)?.name
    }

    init(owner: SchedulingOwner, eventTypeId: String, client: SchedulingClient) {
        self.owner = owner
        self.eventTypeId = eventTypeId
        self.client = client
    }

    // MARK: Lifecycle

    func load() async {
        phase = .loading
        do {
            let pool = try await AssignmentLoader.loadPool(owner: owner, eventTypeId: eventTypeId, client: client)
            let assignedById = Dictionary(pool.assignees.map { ($0.subjectId, $0) }) { first, _ in first }
            picks = pool.members.map { member in
                let assignee = assignedById[member.id]
                return Pick(
                    id: member.id,
                    name: member.name,
                    avatarURL: member.avatarURL,
                    role: member.role,
                    checked: assignee != nil,
                    weight: assignee?.weight ?? 1
                )
            }
            // Members assigned but not in the roster fetch (defensive) — keep them.
            let knownIds = Set(picks.map(\.id))
            for assignee in pool.assignees where !knownIds.contains(assignee.subjectId) {
                picks.append(Pick(
                    id: assignee.subjectId,
                    name: "Member",
                    avatarURL: nil,
                    role: nil,
                    checked: true,
                    weight: assignee.weight ?? 1
                ))
            }
            selectedRule = Self.inferRule(pool.assignees)
            phase = .ready
        } catch let error as SchedulingError {
            phase = .error(error.userMessage ?? "Couldn't load assignment.")
        } catch {
            phase = .error("Couldn't load assignment.")
        }
    }

    // MARK: Editing

    func selectRule(_ rule: Rule) {
        selectedRule = rule
    }

    func toggle(_ id: String) {
        guard let idx = picks.firstIndex(where: { $0.id == id }) else { return }
        picks[idx].checked.toggle()
    }

    func incrementWeight(_ id: String) {
        guard let idx = picks.firstIndex(where: { $0.id == id }) else { return }
        picks[idx].weight = min(picks[idx].weight + 1, 9)
    }

    func decrementWeight(_ id: String) {
        guard let idx = picks.firstIndex(where: { $0.id == id }) else { return }
        picks[idx].weight = max(picks[idx].weight - 1, 1)
    }

    func moveUp(_ id: String) {
        guard let idx = picks.firstIndex(where: { $0.id == id }), idx > 0 else { return }
        picks.swapAt(idx, idx - 1)
    }

    func moveDown(_ id: String) {
        guard let idx = picks.firstIndex(where: { $0.id == id }), idx < picks.count - 1 else { return }
        picks.swapAt(idx, idx + 1)
    }

    // MARK: Save

    func save() async -> Bool {
        guard !doneDisabled else { return false }
        isSaving = true
        saveError = nil
        defer { isSaving = false }

        let checked = picks.filter(\.checked)
        // Sentinel encoding (see header): the rule must survive save→reload,
        // so each rule writes a shape the other two can never produce —
        // balanced marks every row priority -1, priority ranks 1-based, strict
        // stays all-zero. Previously balanced-with-default-weights and
        // single-member priority round-tripped back as "Strict".
        let assignees: [AssigneesRequest.Assignee] = checked.enumerated().map { index, pick in
            switch selectedRule {
            case .balanced:
                AssigneesRequest.Assignee(subjectId: pick.id, subjectType: "user", weight: pick.weight, priority: -1)
            case .priority:
                AssigneesRequest.Assignee(subjectId: pick.id, subjectType: "user", weight: 1, priority: index + 1)
            case .strict:
                AssigneesRequest.Assignee(subjectId: pick.id, subjectType: "user", weight: 1, priority: 0)
            }
        }
        do {
            _ = try await client.request(
                SchedulingEndpoints.setEventTypeAssignees(owner: owner, id: eventTypeId, AssigneesRequest(assignees: assignees)),
                as: AssigneesResponse.self
            )
            return true
        } catch let error as SchedulingError {
            if error.validationDetails.contains(where: { $0.code == "INVALID_ASSIGNEE" }) || error.code == "INVALID_ASSIGNEE" {
                saveError = "One of those members isn't on your team anymore. Refresh and try again."
            } else {
                saveError = error.userMessage ?? "Couldn't save assignment."
            }
            return false
        } catch {
            saveError = "Couldn't save assignment."
            return false
        }
    }

    // MARK: Helpers

    private static func inferRule(_ assignees: [EventTypeAssigneeDTO]) -> Rule {
        guard !assignees.isEmpty else { return .balanced }
        // Sentinel decode (see header). Negative priority = balanced marker;
        // positive = priority rank. Non-default weights also read as balanced
        // so rows saved by older clients (priority 0 + custom weights) keep
        // their rule.
        if assignees.contains(where: { ($0.priority ?? 0) < 0 }) { return .balanced }
        if assignees.contains(where: { ($0.weight ?? 1) != 1 }) { return .balanced }
        if assignees.contains(where: { ($0.priority ?? 0) > 0 }) { return .priority }
        return .strict
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/scheduling/business/RoundRobinAssignmentViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions", "MagicNumber")

package app.pantopus.android.ui.screens.scheduling.business

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.scheduling.AssigneeInput
import app.pantopus.android.data.api.models.scheduling.AssigneesRequest
import app.pantopus.android.data.api.models.scheduling.EventTypeAssigneeDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.auth.AuthRepository
import app.pantopus.android.data.businesses.BusinessTeamRepository
import app.pantopus.android.data.scheduling.SchedulingError
import app.pantopus.android.data.scheduling.SchedulingErrorDecoder
import app.pantopus.android.data.scheduling.SchedulingOwner
import app.pantopus.android.data.scheduling.SchedulingRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * G1 Round-Robin Assignment (Stream A13) — a local bottom sheet hung off a
 * business service. Picks which members take bookings and the fairness rule.
 * Saving REPLACES the whole assignee set via `PUT /event-types/:id/assignees`
 * (`INVALID_ASSIGNEE` surfaced). Mirrors iOS `RoundRobinAssignmentViewModel`
 * + `roundrobin-frames.jsx`.
 *
 * Backend note: there is no dedicated "rule" column — Balanced/Priority/Strict
 * are expressed through the assignees' weight/priority and inferred on reload.
 * The encoding is a lossless sentinel scheme (identical to iOS) so the chosen
 * rule survives a save→reload roundtrip:
 *   · Strict   → weight 1, priority 0            (the neutral shape)
 *   · Priority → weight 1, priority index+1      (1-based rank; >0 marks the rule)
 *   · Balanced → user weights, priority -1       (<0 marks the rule; weights carry the data)
 * (bookingService.pickRoundRobinHost sorts by assigned_count then priority, so
 * a constant -1 across the set never changes the pick; Joi and the DB column
 * both accept any integer priority.) Previously Balanced-with-default-weights
 * and single-member Priority both round-tripped back as "Strict".
 */
@HiltViewModel
class RoundRobinAssignmentViewModel
    @Inject
    constructor(
        private val repo: SchedulingRepository,
        private val businessTeam: BusinessTeamRepository,
        private val auth: AuthRepository,
        private val errors: SchedulingErrorDecoder,
    ) : ViewModel() {
        enum class Rule { Balanced, Priority, Strict }

        data class PickUi(
            val id: String,
            val name: String,
            val role: String?,
            val checked: Boolean,
            val weight: Int,
        )

        sealed interface UiState {
            data object Loading : UiState

            data class Content(
                val rule: Rule,
                val picks: List<PickUi>,
                val saving: Boolean,
            ) : UiState {
                val checkedCount: Int get() = picks.count { it.checked }
                val doneDisabled: Boolean get() = checkedCount == 0 || saving
                val isSingleMember: Boolean get() = checkedCount == 1
                val firstCheckedName: String? get() = picks.firstOrNull { it.checked }?.name
            }

            data class Error(val message: String) : UiState
        }

        sealed interface Event {
            data object Saved : Event

            data class Toast(val message: String) : Event
        }

        private val _state = MutableStateFlow<UiState>(UiState.Loading)
        val state: StateFlow<UiState> = _state.asStateFlow()

        private val _events = Channel<Event>(Channel.BUFFERED)
        val events = _events.receiveAsFlow()

        private var eventTypeId: String = ""
        private var started = false

        fun start(id: String) {
            if (started && eventTypeId == id) return
            started = true
            eventTypeId = id
            load()
        }

        fun load() {
            _state.value = UiState.Loading
            val owner = businessOwner()
            if (owner == null) {
                _state.value = UiState.Error("Switch to a business profile to assign bookings.")
                return
            }
            viewModelScope.launch {
                when (val result = loadAssignmentPool(repo, businessTeam, owner, eventTypeId)) {
                    is NetworkResult.Failure ->
                        _state.value = UiState.Error(errors.decode(result.error).displayMessage("Couldn't load assignment."))
                    is NetworkResult.Success -> {
                        val pool = result.data
                        val assignedById = pool.assignees.associateBy { it.subjectId }
                        val picks =
                            pool.members.map { member ->
                                val assignee = assignedById[member.id]
                                PickUi(
                                    id = member.id,
                                    name = member.name,
                                    role = member.role,
                                    checked = assignee != null,
                                    weight = assignee?.weight ?: 1,
                                )
                            }.toMutableList()
                        // Keep any assignee not present in the roster fetch (defensive).
                        val known = picks.map { it.id }.toSet()
                        pool.assignees.filter { it.subjectId !in known }.forEach {
                            picks.add(PickUi(id = it.subjectId, name = "Member", role = null, checked = true, weight = it.weight ?: 1))
                        }
                        _state.value = UiState.Content(rule = inferRule(pool.assignees), picks = picks, saving = false)
                    }
                }
            }
        }

        // ─── Editing ────────────────────────────────────────────────────────────

        fun selectRule(rule: Rule) = mutate { it.copy(rule = rule) }

        fun toggle(id: String) =
            mutate { content ->
                content.copy(picks = content.picks.map { if (it.id == id) it.copy(checked = !it.checked) else it })
            }

        fun incrementWeight(id: String) =
            mutate { content ->
                content.copy(picks = content.picks.map { if (it.id == id) it.copy(weight = (it.weight + 1).coerceAtMost(9)) else it })
            }

        fun decrementWeight(id: String) =
            mutate { content ->
                content.copy(picks = content.picks.map { if (it.id == id) it.copy(weight = (it.weight - 1).coerceAtLeast(1)) else it })
            }

        fun moveUp(id: String) =
            mutate { content ->
                val idx = content.picks.indexOfFirst { it.id == id }
                if (idx <= 0) {
                    content
                } else {
                    content.copy(picks = content.picks.toMutableList().apply { add(idx - 1, removeAt(idx)) })
                }
            }

        fun moveDown(id: String) =
            mutate { content ->
                val idx = content.picks.indexOfFirst { it.id == id }
                if (idx < 0 || idx >= content.picks.lastIndex) {
                    content
                } else {
                    content.copy(picks = content.picks.toMutableList().apply { add(idx + 1, removeAt(idx)) })
                }
            }

        private inline fun mutate(transform: (UiState.Content) -> UiState.Content) {
            val current = _state.value as? UiState.Content ?: return
            _state.value = transform(current)
        }

        // ─── Save ─────────────────────────────────────────────────────────────

        fun save() {
            val content = _state.value as? UiState.Content ?: return
            if (content.doneDisabled) return
            _state.value = content.copy(saving = true)
            val owner = businessOwner() ?: return
            viewModelScope.launch {
                val checked = content.picks.filter { it.checked }
                // Sentinel encoding (see class doc): each rule writes a shape the
                // other two can never produce — Balanced marks every row with
                // priority -1, Priority ranks 1-based, Strict stays all-zero.
                val assignees =
                    checked.mapIndexed { index, pick ->
                        when (content.rule) {
                            Rule.Balanced ->
                                AssigneeInput(
                                    subjectId = pick.id,
                                    subjectType = SUBJECT_USER,
                                    weight = pick.weight,
                                    priority = BALANCED_PRIORITY_SENTINEL,
                                )
                            Rule.Priority ->
                                AssigneeInput(subjectId = pick.id, subjectType = SUBJECT_USER, weight = 1, priority = index + 1)
                            Rule.Strict -> AssigneeInput(subjectId = pick.id, subjectType = SUBJECT_USER, weight = 1, priority = 0)
                        }
                    }
                when (val result = repo.setAssignees(owner, eventTypeId, AssigneesRequest(assignees))) {
                    is NetworkResult.Success -> _events.send(Event.Saved)
                    is NetworkResult.Failure -> {
                        _state.value = content.copy(saving = false)
                        _events.send(Event.Toast(assigneeErrorMessage(errors.decode(result.error))))
                    }
                }
            }
        }

        // ─── Helpers ────────────────────────────────────────────────────────────

        private fun businessOwner(): SchedulingOwner.Business? =
            (auth.state.value as? AuthRepository.State.SignedIn)?.user?.id?.let { SchedulingOwner.Business(it) }

        // Sentinel decode (see class doc; mirrors iOS `inferRule`). Negative
        // priority = Balanced marker; positive = Priority rank. Non-default
        // weights also read as Balanced so rows saved by older clients
        // (priority 0 + custom weights) keep their rule.
        private fun inferRule(assignees: List<EventTypeAssigneeDto>): Rule =
            when {
                assignees.isEmpty() -> Rule.Balanced
                assignees.any { (it.priority ?: 0) < 0 } -> Rule.Balanced
                assignees.any { (it.weight ?: 1) != 1 } -> Rule.Balanced
                assignees.any { (it.priority ?: 0) > 0 } -> Rule.Priority
                else -> Rule.Strict
            }

        private fun assigneeErrorMessage(error: SchedulingError): String =
            when {
                error is SchedulingError.Generic && error.code == CODE_INVALID_ASSIGNEE ->
                    "One of those members isn't on your team anymore. Refresh and try again."
                error is SchedulingError.Validation && error.details.any { it.code == CODE_INVALID_ASSIGNEE } ->
                    "One of those members isn't on your team anymore. Refresh and try again."
                error is SchedulingError.Generic -> error.message
                else -> "Couldn't save assignment."
            }

        private fun SchedulingError.displayMessage(fallback: String): String =
            when (this) {
                is SchedulingError.Generic -> message
                else -> fallback
            }

        // Internal (not private) so tests can assert against the wire encoding itself
        // rather than re-hardcoding the sentinel and drifting from it.
        internal companion object {
            const val SUBJECT_USER = "user"
            const val CODE_INVALID_ASSIGNEE = "INVALID_ASSIGNEE"

            /** Balanced marker (see class doc) — shared encoding with iOS. */
            const val BALANCED_PRIORITY_SENTINEL = -1
        }
    }
